Ensure scrobbles outside MPD are cached correctly

danielg0 opened the pull request:
When scrobbles of tracks not in the MPD library fail, the duration of the track wasn't stored in the `scrobbles.cache` file as it wasn't present within the `track_info` dictionary like with #6. Similar errors occurred in `make_scrobble` and `now_playing` subroutines as well. In the process, I've also generalised the almost identical code for the creation of parameter lists/scrobbles in all those three functions with `make_scrobble` which takes both the status and track info, as well as an optional `api_sig` for signing the resulting scrobble and any other parameters passed as well.

As this left like five lines of code in the `record_failed_scrobble`, I've deleted it and simply placed its functionality directly where it's needed in the `mpd_watch_track`. I haven't removed it yet, but I was curious as to the need to [check](https://github.com/Berulacks/yams/blob/master/yams/scrobble.py#L463) whether a failed scrobble already exists within `failed_scrobbles` - as each has a timestamp, it would seem this wouldn't be an issue as long as two identical tracks aren't scrobbled in the the same second, which would be a bug in itself.

danielg0 committed:
Prevent crash when reading scrobble cache

Syntax error when duration not present in scrobbe cache, caused by the use of
track_info rather than tracks[i] in elif condition.

danielg0 committed:
Create make_scrobble function

Generalise the creation of parameter lists for last.fm requests or for
storing failed scrobbles. Takes the value from the player status and
song, as well as optional values to place in the returned dict too. If
an api secret is given, sign the scrobble too.

This can reduce the amount of repetitive code in other functions
(now_playing, send_scrobble) and make it easier to add additional
properties as needed in future.

danielg0 committed:
Use make_scrobble function in scrobble functions

Use the new make_scrobble function to replace the parameter creation
functionality in the now_playing and scrobble_track. As very little code
remained in the record_failed_scrobble function after this change, I've
deleted it and simply placed the check for whether the failed scrobble
already exists in the two places it's needed.

danielg0 committed:
Use arbitrary kwargs in make_scrobble()

Rather than specifying all the possible parameters that can be added to
a scrobble, instead take any provided key-value pairs and add it to the
scrobble before signing.

This means that argument names have to match the parameter name in the
last.fm api, which is why the change from session_key to sk is needed.

Berulacks commented:
At a cursory glance looks very solid - I like `make_scrobble` and how it cleans up the code base, thanks! Will give it a thorough check later in the day, should be able to do a proper review around then. 🤘

Berulacks commented on yams/scrobble.py:
```diff
@@ -879,18 +870,34 @@ def mpd_watch_track(client, session, config):
+                                        cache_file_path, failed_scrobbles
+                                    )
                         else:
                             # If we have failed and queued up scrobbles, add this one to the list and try to do them all in one go
-                            record_failed_scrobble(
-                                song, start_time, failed_scrobbles, cache_file_path
+                            failed_scrobble = make_scrobble(
+                                song, status, timestamp=start_time
                             )
+                            if failed_scrobble not in failed_scrobbles:
+                                failed_scrobbles.append(failed_scrobble)
+                                save_failed_scrobbles_to_disk(
+                                    cache_file_path, failed_scrobbles
+                                )
```
I wonder if we could squash these repeating lines into a `record_failed_scrobble` function? The biggest reason for that function was simply to prevent repeating lines and to slightly improve legibility.  Just a minor niggle.

I'll probably re-create that function after the merge and add lines 883-890/896-900 into it - if that's all right with you? (or if you want to do it and I can wait on the PR)

danielg0 replied:
Yeah, you could - that's what was left of the pre-existing `record_failed_scrobble` function before I removed it. The only thing is, I wasn't sure about whether the check for if `failed_scrobble` was in `failed_scrobbles` was needed, as from what I can see, each `failed_scrobble` has a unique timestamp, and so shouldn't match with anything in `failed_scrobbles` anyway (unless the user's been messing about with their clock). If so, both 883-890 and 896-900 could be reduced to:

```python
failed_scrobbles.append(make_scrobble(
    song, status, timestamp=start_time
)
save_failed_scrobbles_to_disk(
    cache_file_path, failed_scrobbles
)
```

Which seemed a little trivial to make an entire function out of. However, if the check is necessary, I'd be fine with you recreating that function, it'd make that section clearer.

Berulacks commented on yams/scrobble.py:
```diff
@@ -428,83 +463,39 @@ def scrobble_tracks(tracks, url, api_key, api_secret, session_key):
-
-    parameters["api_sig"] = sign_signature(parameters, api_secret)
+    logger.info("Scrobbling!")
+    parameters = make_scrobble(
```
Thanks for this - it's much cleaner.

Berulacks commented on yams/scrobble.py:
```diff
@@ -384,7 +419,7 @@ def scrobble_tracks(tracks, url, api_key, api_secret, session_key):
         # We do this for older clients, such as mopidy, that use the "time" variable to send duration data
         # Which is deprecated according to the mpd protocol. Oh well. Bad mopidy, bad.
-        elif "time" in track_info:
+        elif "time" in tracks[i]:
```
I don't think this is possible? As per line 353, you don't actually ever place a value into the "time" key, but into "duration". I recommend we just clear  this line entirely as the whole "checking for time vs duration" thing is now handled in `make_scrobble`

Edit: Oh FFS I had to hit "Submit Review" to get this comment out? My bad 😅

danielg0 replied:
Yeah, you're totally right. Gimmie a moment, I'll just remove that whole segment.

danielg0 committed:
Remove check for time in scrobble_tracks

Unnecessary as the check for time in the absence of duration is
performed in the make_scrobble function, which is used for all scrobbles
added to failed_scrobbles. Thanks @Berulacks

Berulacks commented:
LGTM, thanks again!
